### PR TITLE
Make detours common for dup, dup2, dup3, & fcntl

### LINKED ISSUE #258: Support dup for file-ops
Support "Duplicating a file descriptor" commands for file-ops and not only for sockets.

Syscalls like `dup`, `dup2`, `dup3` and `fcntl` can duplicate duplicate a file descriptor, if the fd is one that mirrord manages then post duplication the new fd still should be mapped to the remote-fd

### PR DESCRIPTION
**Pull request by infiniteregrets**
Will close #258 and #259

### CODE REVIEW

File: mirrord-layer/src/common.rs [hunk 1]
```diff
@@ -29,3 +38,172 @@ pub enum HookMessage {
     File(HookMessageFile),
     GetAddrInfoHook(GetAddrInfoHook),
 }
+
+mod common_ops {
```
Comment (meowjesty) [inline mirrord-layer/src/common.rs:44]: I suggest moving these modules under a separate thing, this `common` module name is already not very helpful. Maybe have a `common_detour` folder/module, and inside have the `hooks` and `ops`, so we access those as `common_detour::hooks` and `common_detour::ops`.

File: mirrord-layer/src/common.rs [hunk 2]
```diff
@@ -29,3 +40,172 @@ pub enum HookMessage {
...
+    use std::{os::unix::prelude::RawFd, sync::OnceLock};
+
+    use libc::c_int;
+
```
Comment (meowjesty) [inline mirrord-layer/src/common.rs:48]: Why do you need this here? Can't you use the already global `ENABLED_FILE_OPS` we have?

**Review by meowjesty — COMMENTED** [review]
Some initial thoughts on the code.

File: mirrord-layer/src/common_hooks.rs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        trace!("fcntl_detour -> fd {:#?} | cmd {:#?}", fd, cmd);
+
+        let arg = arg.arg::<usize>();
+        let fcntl_result = FN_FCNTL(fd, cmd, arg);
+        let guard = crate::DetourGuard::new();
+        if guard.is_none() {
+            return fcntl_result;
+        }
+
+        if fcntl_result == -1 {
```
Comment (eyalb181) [inline mirrord-layer/src/common_hooks.rs:59]: Why not combine it with the above condition? Or even do it before, so you don't needlessly initialize a DetourGuard if fcntl_result is -1 anyway

File: mirrord-layer/src/common_hooks.rs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    use crate::error::LayerError;
+
+    /// https://github.com/metalbear-co/mirrord/issues/184
+    #[hook_fn]
+    pub(super) unsafe extern "C" fn fcntl_detour(fd: c_int, cmd: c_int, mut arg: ...) -> c_int {
+        trace!("fcntl_detour -> fd {:#?} | cmd {:#?}", fd, cmd);
+
+        let arg = arg.arg::<usize>();
+        let fcntl_result = FN_FCNTL(fd, cmd, arg);
+        let guard = crate::DetourGuard::new();
```
Comment (eyalb181) [inline mirrord-layer/src/common_hooks.rs:54]: Why is DetourGuard used here but not in e.g. dup?

**Comment by infiniteregrets** [comment]
stale pr, closing this will open a new one once mvp is done

**Pull request closed without merging**